### PR TITLE
Allow creating workbooks with multiple sheets

### PR DESCRIPTION
**Pull request by dpetranek**
Modifies the create-workbook function without breaking backwards
compatibly. The original arity of [sheetname data] is preserved,
but I added an arity of [sheet-vec] that takes a vector of
alternating sheetnames and sheetdata.

### CODE REVIEW

**Comment by dpetranek** [comment]
Currently, not all the tests are passing - specifically checking row numbers. I think it's a problem with how I wrote the test instead of the functionality, though.

File: src/dk/ative/docjure/spreadsheet.clj [hunk 1]
```diff
@@ -272,22 +272,32 @@
...
+         sheet    (add-sheet! workbook sheet-name)]
+     (add-rows! sheet data)
+     workbook))
+  
+  ([sheet-vec]
+   (let [workbook (XSSFWorkbook.)
+        sheet-groups (partition 2 2 sheet-vec)
+        sheets (map #(conj [] (add-sheet! workbook (first %)) (second %)) sheet-groups)]
+    (doseq [sheet sheets]
+      (add-rows! (first sheet) (second sheet)))
```
Comment (slipset) [inline src/dk/ative/docjure/spreadsheet.clj]: `(apply add-rows! sheet)`

**Comment by slipset** [comment]
@dpetranek I'd be very interested in this functionality :)

**Comment by madstap** [comment]
@dpetranek Nice, I was just thinking about how the create-workbook function should have this and then I saw your PR.

The version I pictured uses variable arity though, `[sheet-name data & name-data-pairs]`, the rationale being that sheet names and data are key value pairs and create-workbook should work like `hash-map` or `assoc`. Why are the pairs wrapped in a vector?

File: src/dk/ative/docjure/spreadsheet.clj [hunk 2]
```diff
@@ -272,22 +272,32 @@
...
+  associated sheetdata [SheetName1, SheetData1, SheetName2, SheetData2, ...]
+
+  (create-workbook [\"SheetName1\" [[\"FirstSheet A1\"][\"FirstSheet A2\"]]
+                    \"SheetName2\" [[\"SecondSheet A1\"][\"SecondSheet A2\"]]])"
+  ([sheet-name data]
+   (let [workbook (XSSFWorkbook.)
+         sheet    (add-sheet! workbook sheet-name)]
+     (add-rows! sheet data)
+     workbook))
+
```
Comment (madstap) [inline src/dk/ative/docjure/spreadsheet.clj]: I had some trouble following your definition of sheets, so I wrote an implementation of a cleaned up version (clean according to me, of course).

```clojure
([sheet-name data & name-data-pairs]
 ;; Incomplete pairs should not be allowed.
 {:pre [(even? (count name-data-pairs))]}
 ;; Use the two arity version of the function to make the initial workbook
 (let [workbook (create-workbook sheet-name data)]
   ;; Then iterate through the pairs adding sheets and rows
   (doseq [[s-name data] (partition 2 name-data-pairs)]
     (-> workbook
         (add-sheet! s-name)
         (add-rows! data)))
   workbook))
```


Single arity version
```clojure
([sheet-vec]
 ;; Incomplete pairs should not be allowed.
 {:pre [(even? (count sheet-vec))]}
 ;; Use the two arity version of the function to make the initial workbook
 (let [workbook (create-workbook (first sheet-vec) (second sheet-vec))]
   ;; Then iterate through the pairs adding sheets and rows
   (doseq [[s-name data] (partition 2 (drop 2 sheet-vec))]
     (-> workbook
         (add-sheet! s-name)
         (add-rows! data)))
   workbook))
```

File: src/dk/ative/docjure/spreadsheet.clj [hunk 3]
```diff
@@ -272,22 +272,32 @@
...
+     (add-rows! sheet data)
+     workbook))
+  
+  ([sheet-vec]
+   (let [workbook (XSSFWorkbook.)
+        sheet-groups (partition 2 2 sheet-vec)
+        sheets (map #(conj [] (add-sheet! workbook (first %)) (second %)) sheet-groups)]
+    (doseq [sheet sheets]
+      (add-rows! (first sheet) (second sheet)))
+    workbook)))
```
Comment (madstap) [inline src/dk/ative/docjure/spreadsheet.clj]: I had some trouble following your definition of sheets, so I wrote an implementation of a cleaned up version (clean according to me, of course).

```clojure
([sheet-name data & name-data-pairs]
 ;; Incomplete pairs should not be allowed.
 {:pre [(even? (count name-data-pairs))]}
 ;; Use the two arity version of the function to make the initial workbook
 (let [workbook (create-workbook sheet-name data)]
   ;; Then iterate through the pairs adding sheets and rows
   (doseq [[s-name data] (partition 2 name-data-pairs)]
     (-> workbook
         (add-sheet! s-name)
         (add-rows! data)))
   workbook))
```


Single arity version
```clojure
([sheet-vec]
 ;; Incomplete pairs should not be allowed.
 {:pre [(even? (count sheet-vec))]}
 ;; Use the two arity version of the function to make the initial workbook
 (let [workbook (create-workbook (first sheet-vec) (second sheet-vec))]
   ;; Then iterate through the pairs adding sheets and rows
   (doseq [[s-name data] (partition 2 (drop 2 sheet-vec))]
     (-> workbook
         (add-sheet! s-name)
         (add-rows! data)))
   workbook))
```

File: README.md [hunk 4]
```diff
@@ -82,6 +82,28 @@ to the top header row, then save the spreadsheet.
...
+### Example: Create a workbook with multiple sheets
+This example creates a spreadsheet with multiple sheets. Instead of just passing a
+sheet name and a dataset, we pass a vector of the form [SheetName1, SheetData1,
+SheetName2, SheetData2, ...].
+
+```clj
+(use 'dk.ative.docjure.spreadsheet)
+
+;; Create a spreadsheet and save it
+(let [wb (create-workbook ["Squares"
```
Comment (slipset) [inline README.md]: Personally, I'm more fond of 
`{:name "Squares
  :data [..]
}`
since I find tuples harder to read than maps

**Comment by dpetranek** [comment]
@madstap I like your idea of using variable arity, I think it's much more intuitive for users to have consistent syntax for this function between the single-sheet and multiple sheet versions. And your cleaned up version is much less hacky than mine was, and a distinct improvement for readability, so I'm going with that.

I also like the `:pre` test to weed out unmatched pairs. I initially decided to be forgiving, and just allow it to create a blank sheet because I sometimes need to do that. But I can just pass `nil `if I need a blank sheet, and I think it's better to check our expectations at the door.

I've committed the new version with cleaned up tests, everything is passing now.

**Comment by dpetranek** [comment]
@mjul Hello Martin, could you take a look at this? @madstap helped me clean up the implementation to simplify and I think it stands pretty well on its own now. 

The tests are based on those for the `create-workbook` function and they are all passing. 

I'm about to start a new project and I'd like to see if we can get this merged before then. Thanks!

**Comment by dpetranek** [comment]
I had commented out the pomegranate dependency for reasons that are now lost to the mists of time, but I just uncommented them and re-ran the tests, everything is still good.

**Comment by mjul** [comment]
Looks good. Please add your names as contributors in the Readme and you are good to go.

**Comment by dpetranek** [comment]
I added mine. Your turn @madstap.

**Comment by madstap** [comment]
Could you just add it?
`[Aleksander Madland Stapnes](https://github.com/madstap) (madstap)`
Or should I make a pull request?

**Comment by dpetranek** [comment]
@madstap I added you.
@mjul Looks like we're good to go. Thanks!

**Comment by dparis** [comment]
Hello all! Any chance this could be merged in the next day or so? If not, I can just use @dpetranek's branch directly.

Thanks for all your work on this library!

**Comment by mjul** [comment]
Thanks for the great work!

**Comment by mjul** [comment]
Hi Dylan

I was just merged and published as the latest snapshot build to Clojars.
Have fun with it and have a nice day, and if you have some ideas for improvements, don’t hesitate to send a pull request ;-)

Cheers,
Martin

**Comment by mjul** [comment]
This is now available in the 1.11-SNAPSHOT build on Clojars:

https://clojars.org/dk.ative/docjure/versions/1.11.0-SNAPSHOT

**Comment by dparis** [comment]
@mjul Thank you very much, I look forward to using it! And yes, I'll be sure to send a PR if we come up with any fixes/improvements.

@dpetranek @madstap Thanks as well for your work on contributing this feature!

**Pull request merged**